### PR TITLE
ONE: The Eternal Wanderer + support + tidy related cards

### LINKED ISSUE #2387: The Eternal Wanderer (& w_2_2_samurai_double_strike token)
Working on refactoring a few of the Global Rules keywords as well while working on this:

Image

### PR DESCRIPTION
**Pull request by Northmoc**
Closes #2387

### CODE REVIEW

File: forge-game/src/main/java/forge/game/combat/GlobalAttackRestrictions.java [hunk 1]
```diff
@@ -133,7 +134,7 @@ public static GlobalAttackRestrictions getGlobalRestrictions(final Player attack
         }
 
         for (final GameEntity defender : possibleDefenders) {
-            final int defMax = getMaxAttackTo(defender);
+            final int defMax = StaticAbilityAttackRestrict.attackRestrictNum(defender);
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/combat/GlobalAttackRestrictions.java:137]: The KW and the global rule above there should be mergeable too?

File: forge-game/src/main/java/forge/game/staticability/StaticAbilityAttackRestrict.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+                    return AbilityUtils.calculateAmount(stAb.getHostCard(),
+                            stAb.getParamOrDefault("MaxAttackers", "1"), stAb);
+                }
+            }
+        }
+        return -1;
+    }
+
+    static public boolean attackRestrict(StaticAbility stAb, GameEntity defender) {
+        if (!stAb.matchesValidParam("DefenderDefined", defender)) {
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/staticability/StaticAbilityAttackRestrict.java:30]: sounds a bit confusing bc it's not defined syntax

ValidDefender perhaps?

File: forge-game/src/main/java/forge/game/combat/GlobalAttackRestrictions.java (hunk 1 above)
Comment (Northmoc) [reply forge-game/src/main/java/forge/game/combat/GlobalAttackRestrictions.java:137]: Worth a shot!

File: forge-gui/res/cardsfolder/upcoming/the_eternal_wanderer.txt [hunk 3]
```diff
@@ -0,0 +1,17 @@
+Name:The Eternal Wanderer
+ManaCost:4 W W
+Types:Legendary Planeswalker
+Loyalty:5
+S:Mode$ AttackRestrict | MaxAttackers$ 1 | DefenderDefined$ Card.Self | Description$ No more than one creature can attack CARDNAME each combat.
+A:AB$ ChangeZone | Cost$ AddCounter<1/LOYALTY> | Planeswalker$ True | Origin$ Battlefield | Destination$ Exile | ValidTgts$ Artifact,Creature | TargetMin$ 0 | TargetMax$ 1 | TgtPrompt$ Select up to one target artifact or creature | RememberChanged$ True | SubAbility$ DBDelTrig | SpellDescription$ Exile up to one target artifact or creature. Return that card to the battlefield under its owner's control at the beginning of that player's next end step.
+SVar:DBDelTrig:DB$ DelayedTrigger | Mode$ Phase | Phase$ End of Turn | DelayedTriggerDefinedPlayer$ TargetedOwner | Execute$ TrigReturn | RememberObjects$ RememberedLKI | SubAbility$ DBCleanup | TriggerDescription$ Return that card to the battlefield under its owner's control at the beginning of that player's next end step. | StackDescription$ Return that card to the battlefield under its owner's control at the beginning of that player's next end step.
```
Comment (tool4ever) [inline forge-gui/res/cardsfolder/upcoming/the_eternal_wanderer.txt:7]: This should be done with ValidPlayer

I'll add PlayerProperty IsTriggerRemembered in my next PR

Comment (Northmoc) [reply forge-gui/res/cardsfolder/upcoming/the_eternal_wanderer.txt:7]: ✅

File: forge-game/src/main/java/forge/game/combat/GlobalAttackRestrictions.java (hunk 1 above)
Comment (Northmoc) [reply forge-game/src/main/java/forge/game/combat/GlobalAttackRestrictions.java:115]: ✅

File: forge-game/src/main/java/forge/game/combat/GlobalAttackRestrictions.java [hunk 4]
```diff
@@ -107,33 +106,13 @@ public MapToAmount<GameEntity> getDefenderTooMany() {
...
 
-        /* if (game.getStaticEffects().getGlobalRuleChange(GlobalRuleChange.onlyOneAttackerATurn)) {
-            if (!attackingPlayer.getAttackedWithCreatureThisTurn().isEmpty()) {
-                max = 0;
-            } else {
-                max = 1;
-            }
-        } */
-
-        if (max == -1 && game.getStaticEffects().getGlobalRuleChange(GlobalRuleChange.onlyOneAttackerACombat)) {
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/combat/GlobalAttackRestrictions.java:122]: Looks good!
Just need to also remove this from GlobalRuleChange

File: forge-game/src/main/java/forge/game/staticability/StaticAbilityAttackRestrict.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    static public int attackRestrictNum(GameEntity defender) {
+        final Game game = defender.getGame();
+        for (final Card ca : game.getCardsIn(ZoneType.STATIC_ABILITIES_SOURCE_ZONES)) {
+            for (final StaticAbility stAb : ca.getStaticAbilities()) {
+                if (!stAb.getParam("Mode").equals(MODE) || stAb.isSuppressed() || !stAb.checkConditions()) {
+                    continue;
+                }
+                if (attackRestrict(stAb, defender)) {
+                    return AbilityUtils.calculateAmount(stAb.getHostCard(),
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/staticability/StaticAbilityAttackRestrict.java:39]: You also need to iterate over all here like above

Else if you have both Mirri + Crawlspace you might be in trouble

Comment (Northmoc) [reply forge-game/src/main/java/forge/game/staticability/StaticAbilityAttackRestrict.java:39]: Good point

File: forge-game/src/main/java/forge/game/staticability/StaticAbilityAttackRestrict.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            }
+        }
+        return max < Integer.MAX_VALUE ? max : -1;
+    }
+
+    static public int attackRestrictNum(GameEntity defender) {
+        final Game game = defender.getGame();
+        for (final Card ca : game.getCardsIn(ZoneType.STATIC_ABILITIES_SOURCE_ZONES)) {
+            for (final StaticAbility stAb : ca.getStaticAbilities()) {
+                if (!stAb.getParam("Mode").equals(MODE) || stAb.isSuppressed() || !stAb.checkConditions()) {
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/staticability/StaticAbilityAttackRestrict.java]: You need to check also check if it doesn't have ValidDefender or the others would be counted too

**Review by tool4ever — APPROVED** [review]
nice work

**Pull request merged**